File: 지연(이)/12월18일/PGS 혼자서 하는 틱택톡.java
import java.util.*;

class Solution{
    static int O = 0;
    static int X = 0;
    static int O_line = 0;
    static int X_line = 0;

    static int[][] maps = new int[3][3];

    public void init(String[] board){
        for(int i=0; i<3;i++){
            for(int j=0;j<3;j++){
                char temp = board[i].charAt(j);
                if (temp=='O'){
                    maps[i][j] = 1;
                    O++;
                }else if (temp=='X'){
                    maps[i][j]=-1;
                    X++;
                }
            }
        }
    }

    public int check(int key){
        for (int i=0;i<3;i++){
            if (key==maps[i][0] && maps[i][0]==maps[i][1] && maps[i][1]==maps[i][2]) return 1;
            if (key==maps[0][i] && maps[0][i]==maps[1][i] && maps[1][i]==maps[2][i]) return 1;
        }
        if (key==maps[0][0] && maps[0][0]==maps[1][1] && maps[1][1]==maps[2][2]) return 1;
        if (key==maps[2][0] && maps[2][0]==maps[1][1] && maps[1][1]==maps[0][2]) return 1;
        return 0;
    }

    public int solution(String[] board){
        int answer = 1;
        init(board);
        O_line = check(1);
        X_line = check(-1);

        if (O>=X){
            if (O>X+1) answer = 0;
            else if (O_line==1 && X_line==1) answer = 0;
            else if (O_line==1 && O==X) answer = 0;
            else if (X_line==1 && O>X) answer = 0;
            else answer = 1;
        }else answer = 0;

        return answer;
    }
}

File: 지연(이)/12월18일/PGS 혼자서 하는 틱택톡.py
# 1. 3*3 전체 탐색, 모든 경우의 수를 대비 해도 될듯. 1이 되는 경우
# 2. O count X count, O line count, X line count 이걸로 비교 해서 파악할 수도 있을듯

def solution(board):
    O, X, O_line, X_line = 0,0,0,0
    for i in range(3):
        # 가로
        if board[i][0]==board[i][1]==board[i][2]=='O':
            O_line+=1
        elif board[i][0]==board[i][1]==board[i][2]=='X':
            X_line+=1

        # 세로
        if board[0][i] == board[1][i] == board[2][i] == 'O':
            O_line+=1
        elif board[0][i] == board[1][i] == board[2][i] == 'X':
            X_line+=1

        # 대각선
        if board[0][0]==board[1][1]==board[2][2]=='O':
            O_line+=1
        elif board[0][0]==board[1][1]==board[2][2]=='X':
            X_line+=1

        O += board[i].count('O')
        X += board[i].count('X')

    print(O,X, O_line, X_line)
    #
    if X==O:
        if O_line:
            return 0
        if X>4:
            return 0

        if X_line:
            if (X==3 or X==4) and X_line==1:
                return 1
            else:
                return 0
        else:
            return 1

    elif O==X+1:
        if X_line:
            return 0
        if O>5:
            return 0

        if O_line==0:
            return 1
        else:
            if (O_line==1 or (O_line==2 and O==5)):
                return 1
            else:
                return 0
    else:
        return 0

print(solution(["O.X", ".O.", "..X"])) # 1
print(solution(["OOO", "...", "XXX"])) # 0
print(solution(["...", ".X.", "..."])) # 0
print(solution(["...", "...", "..."])) # 1
print(solution(["OXX", "OOX", "OXO"])) # 1
